File: vector-tracer.cpp
#include <iostream>
#include <math.h>
#include <map>
#include <vector>

using namespace std; 

class Node {
	public:
		bool is_hash = true;
		bool is_string = false;
		map <string, Node*> hash_node;
		vector <Node*> vector_node;
		vector <string> strings;
		
		Node () {
		};
};




class VectorTracer {
	public:
		Node *node = new Node;
		string digit = "";
		int digit_step = 0;
		
		VectorTracer() {
		}
		
		void dump (Node *n = NULL) {
			if (n == NULL) n = this->node;
			bool h = n->is_hash;
			
			if (h) {
				cout << "{";
				auto it = n->hash_node.begin();
				for (it; it != n->hash_node.end() ; it++) {
					string key = it->first;
					if (key != "") {
						cout << key << " => " << endl;
						Node *value = it->second;
						this->dump(value);
					}
				}
				cout << "}";
			} else {
				if (n->strings.size() > 0) {
					cout << "[";
					auto it = n->strings.begin();
					for (it; it != n->strings.end() ; it++) {
						cout << (*it) << "\t";
					}
					cout << "]";
					cout << endl;
				} else {
					cout << "[";
					if (n->vector_node.size() > 0) {
						auto it = n->vector_node.begin();
						for (it; it != n->vector_node.end() ; it++) {
							this->dump(*it);
						}
					}
					cout << "]";
				}
			}
		}
		
		void parse(string str) {
			this->digit = "";
			this->node = new Node;
			str = this->prepare_multi_and_div(str);
			Node *n = this->_parse(str);
			
			//this->_depack(
				
			//);
			this->node = n;
			return;
		}
				
		float trace (Node *node_param = NULL) {
			Node *node = NULL;
			if (node_param == NULL) {
				node = this->node;
			} else {
				node = node_param;
			}
			
			if (node->is_hash) {
				auto it = node->hash_node.begin();
				string key = it->first;
				Node *value = node->hash_node[key];	
				if (key == "sin" || key == "cos") {
					float a = this->trace(value);
					if (key.compare("sin")) return sin(a);
					if (key.compare("cos")) return cos(a);
				} else if (key == "+" || key == "-" || key == "*" || key == "/" || key == "**") {
					float a, b;
					if (value->is_string) {
						vector<string> s = value->strings;
						a = stof(s[0]);
						b = stof(s[1]);
					} else {
						vector<Node*> v = value->vector_node;
						auto i = v.begin();
						a = this->trace(i[0]);
						b = this->trace(i[1]);					
					}


					if (key == "+") return a + b;
					if (key == "-") return a - b;
					if (key == "*") return a * b;
					if (key == "/") return a / b;
					if (key == "**") return pow(a, b);
				}
			} 
			return 0;
		}
		
		string prepare_multi_and_div (string s) {
			int l = s.length();

			for (int i = 0 ; i < l ; i ++) {
				char c = s[i];
				
				if (c == '*' || c == '/') {
					bool p = false;
					if (s[i+1] == '*') p = true ;
					
					// Go back
					for (int j = i - 1; j >= 0 ; j --) {
						if ((j == 0) || s[j] == '*' || s[j] == '/' || s[j] == '+' || s[j] == '-') {
							// Setup brackets
							// Go backward
							for (int k = j ; k >= 0 ; k --) {
								if (s[k] == '*' || s[k] == '/' || s[k] == '+' || s[k] == '-') {
									string s1 = s.substr(0, k+1);
									s = s1 + '(' + s.substr(k+1);
									i ++;
									j ++;
									k ++;
									l ++;
									break;
								}
																
								if (k == 0) {
									s.insert(0, 1, '(');
									
									i ++;
									j ++;
									k ++;
									l ++;
									break;
								} 
							}
	
							// Go forward
							if (p) i ++;
							for (int k = i + 1 ; k < l ; k ++) {
								if (s[k] == '*' || s[k] == '/' || s[k] == '+' || s[k] == '-') {
									string s1 = s.substr(0, k);
									s = s1 + ')' + s.substr(k);
									i ++;
									j ++;
									k ++;
									l ++;
									break;
								}
								if (k == l - 1) {
									
									s += ')';
									i ++;
									j ++;
									k ++;
									l ++;
									break;
								} 

							}

							break;
							
						}					
					}

				}
			}
			return s;
		}
		
		Node * _parse (string s) {
		
		     cout << "Parse string: " + s << endl;
			Node * node = new Node;
			int sl = s.length();

			string function = "";
			vector <string> expression;
			string func = "";
			
			
			for (int i = 0; i < sl; i++) {
				cout << "i=" << i << endl;
				cout << "dump: " << endl;
				this->dump();
				char c = s[i];
				if (c == '+' || c == '-' || c == '*' || c == '/') {
					bool p = false;
					string op = ""+c;
					if (c == '*' && s[i+1] == '*') {
						p = true;
						op = "**";
						i ++;
					}

					if (this->digit != "") {
						expression.push_back(this->digit);
						this->digit = "";
					}
					
					if (node->is_hash == false && node->vector_node.size()) {
						Node *n = new Node;
						n->is_hash = true;
						n->hash_node[op] = node;
						node = n;
					} else if (node->is_hash == true) {
						Node *n = new Node;
						n->is_hash = true;
						n->hash_node[op] = node;
						node = n;
					} else {
						if (c == '+' || c == '-') {
							Node * n = this->_parse_digit(s.substr(i,sl -i));
							node->is_hash = false;
							node->vector_node.push_back(n);
							this->digit = "";
							i = i + this->digit_step;
							continue;							
						}
					}

					int j;
					string buff = "";
					int cnt = 0;
					
					for (j = i + 1; j < sl ; j ++) {
						char o = s[j];
						if (o == '(') {
							cnt ++;
						}
						if (o == ')') {
							cnt --;
						} 
						
						buff += o;
						
						if (cnt < 0 || j == (sl-1)) {

							if (node->hash_node[op]->is_hash) {
								Node * t = new Node;
								t->is_hash = false;
								t->vector_node.push_back(node->hash_node[op]);
								t->vector_node.push_back(this->_parse(buff));
								Node * t2 = new Node;
								t2->is_hash = true;
								t2->hash_node[op] = t;
								node = t2;
							} else {
								// push @{$node->{$op}}, $self->_parse ($buff);
							}
							break;
						} 
					}
					i = j;
				} else if (c == '(') {
					int j;
					string buff = "";
					for (int j = i + 0; j < sl ; j ++) {
						int o = s[j];
						buff += o;
						int cnt = 0;
						if (o == '(') {
							cnt ++;
							continue;
						}
						if (o == ')' || j == sl - 1) {
							cnt --;
							if (cnt == -1) {
								node->vector_node.push_back(this->_parse(buff.substr(1)));
								break;
							}
						}
					}
					i = j;
					function = "";
					continue;
				} else if (std::isalpha(c)) { // fuction processiong
					int j;
					int func_end = 0;
					string buff = "";
					
					for (j = i; j < sl ; j ++) {
						
						char o = s[j];
						if (o == '(') {
							func_end = 1;
							int j2;
							
							int cnt = 1;
							for (j2 = j + 1; j2 < sl ; j2 ++) {
								char o = s[j2];
								buff += o;
								if (o == '(') {
									cnt ++;
									continue;
								}
								if (o == ')') {
									cnt --;
									if (cnt == 0) {
										break;
									}
								}
								
							}
							j = j2;
							break;
						} else if (func_end == 0) {
							func += o;
						}
						
						
					}
					i = j;	
					Node * n = new Node;
					n->hash_node[func] = this->_parse(buff);
					node->vector_node.push_back(n);
				} else if (isdigit(c) || c == '.') {
					cout << "s.substr(i,sl-i) = " << s.substr(i,sl-i) << endl;
					this->digit_step = 0;
					Node * n = this->_parse_digit(s.substr(i,sl-i));
					i = this->digit_step + i;
					this->dump(n);
					cout << "this->digit_step=" << this->digit_step << endl;
					
					node->vector_node.push_back(n);
				}			
			}
	
			return node;
		}
		

		Node * _parse_digit (string s) {
			int i = 0;
			char c = s[0];
			
			Node * node = new Node;
			this->digit += c;
			
			int j = 0;
			string buff = this->digit;
			bool had_non_digit = false;
			
			for (j = i + 1; j < s.length() ; j ++) {
				char o = s[j];
				if (isdigit(o) || o == '.') {
					buff += o;
				} else {
					had_non_digit = true;
					break;
				}
			}
			
			if (had_non_digit) {
				j --;
				this->digit = buff;
				node->strings.push_back(buff);
				node->is_string = true;
				node->is_hash = false;
			} else {
				this->digit = buff;
				node->strings.push_back(buff);
				node->is_string = true;
				node->is_hash = false;
			}
			this->digit_step = j;
			return node;
		}	
				
		Node * _depack (Node *node) {
			if (node->is_hash) {
				auto it = node->hash_node.begin();
				string key = it->first;
				Node *value = it->second;
				
				Node * node = new Node;
				// node->is_hash = true;

				map <string, Node*> n;
				n[key] = this->_depack (value);
				node->hash_node = n;
			} else {
				auto v = node->vector_node;
				if (v.size() == 1) {
					auto n = node[0];
					return this->_depack(&n);
				} else if (v.size() == 0) {
					return node;
				} else {
					for (int i = 0; i < v.size() ; i++) {
						Node n = *(v[i]);
						v[i] = this->_depack(&n);
					}
				}
			}
			
			return node;
		}
};

int main () {
	VectorTracer *tracer = new VectorTracer;
	string math = "123+22";
	tracer->parse(math);
	tracer->dump();
	cout << tracer->trace() << "\n";
}

File: vector-tracer.pl
#!/usr/bin/perl

# This deal I was coded on C++ when I was study in Moscow, try to remember this stuff...

use strict;
use warnings;

use 5.022;
use Data::Dumper;

my $math = "2-2*cos(10)*10";

my $tracer = new VectorTracer(debug => 0);
my $node = $tracer->parse($math);
say $tracer->trace();

package VectorTracer;

use strict;
use warnings;
use experimental 'smartmatch';

use Data::Dumper;

sub new {
	my $class = shift;
	my %opts  = (
		debug => 0,
		@_,
	);
	
	my $self = bless {
		node	=> undef,
		digit 	=> '',
		debug   => $opts{debug},
	}, $class;
	
	return $self;
}


sub debug {
	my $self = shift;
	my $msg  = shift;
	
	if ($self->{debug}) {
		warn '['.scalar(localtime).'] '. $msg . "\n";
	}
	return;
}

sub trace {
	my $self = shift;
	my $node = shift // $self->{node};
	
	if (ref $node eq 'HASH') {
		my ($key, $value) = each %$node;
		if ($key ~~ [qw/sin cos/]) {
			my $a = $self->trace($value);
			return sin($a) if ($key eq 'sin');
			return cos($a) if ($key eq 'cos');
		} elsif ($key ~~ ['+', '-', '*', '/']) {
			my ($a, $b) = @$value;
			$a = $self->trace($a);
			$b = $self->trace($b);
			return $a + $b if ($key eq '+');
			return $a - $b if ($key eq '-');
			return $a * $b if ($key eq '*');
			return $a / $b if ($key eq '/');
		}
	} elsif (ref $node eq 'ARRAY') {
		
	}
	return int $node;
}

# Hack method for fix math priority: setup brackets to multi and div
sub prepare_multi_and_div {
	my $self = shift;
	my $str = shift;

	my @s = split //, $str;
	my $l = scalar (@s);
		
	for (my $i = 0 ; $i < $l ; $i ++) {
		my $c = $s[$i];
		if ($c ~~ ['*', '/']) {
			# Go back
			for (my $j = $i - 1; $j >= 0 ; $j --) {
				my $cnt = 0;
				if ($j eq ')') {
					$cnt ++;
					next;
				} 
				if ($j eq '(') {
					$cnt --;
				}
				last if $cnt == -1;
				if (($j == 0) || ($c ~~ ['+','-','/','*'])) {
					# Setup brackets
					# Go backward
					for (my $k = $j -1 ; $k >= 0 ; $k --) {
						if ($k == 0) {
							@s = ('(', @s);
							$i ++;
							$j ++;
							$k ++;
							$l ++;
							last;
						} 
						
						if ($s[$k] ~~ ['+','-','*','/']) {
							my @s1 = splice(@s, 0, $k+1);
							@s = (@s1, '(', @s);
							$i ++;
							$j ++;
							$k ++;
							$l ++;
							last;
						}
					}
					# Go forward
					for (my $k = $i + 1 ; $k < $l ; $k ++) {
						if ($k == $l - 1) {
							push @s, ')';
							$i ++;
							$j ++;
							$k ++;
							$l ++;
							last;
						} 
						if ($s[$k] ~~ ['+','-','*','/']) {
							my @s1 = splice(@s, 0, $k);
							@s = (@s1, ')', @s);
							$i ++;
							$j ++;
							$k ++;
							$l ++;
							last;
						}
					}
					
					last;
				}
			}
		}
	}
	
	return (scalar (@s) >= length ($str)) ? join '', @s : $str;
}


sub parse {
	my $self = shift;
	my $str  = shift;
	$str = $self->prepare_multi_and_div($str);
	my $node = $self->_depack($self->_parse ($str));
	$self->{node} = $node;
	return $node;
}

sub _depack {
	my $self = shift;
	my $node = shift;
	if (ref $node eq 'HASH') {
		my ($key, $value) = each %$node;
		$node = {$key => $self->_depack ($value)};
	} elsif (ref $node eq 'ARRAY') {
		if (scalar (@$node) == 1) {
			$node = $node->[0];
			return $self->_depack($node) ;
		} else {
			for (@$node) {
				$_ = $self->_depack ($_);
			}
		}
	} 
	return $node;
}

sub _parse {
	my $self = shift;
	my $str  = shift;
	my $node = [];
	
	$self->debug("Parse $str");
	
	my @s = split //, $str;
	my $sl = scalar (@s);
	
	my $function = '';
	my @expression = ();
	
	my $value = {};
	
	my $func;
	
	for (my $i = 0; $i < @s; $i++) {
		my $c = $s[$i];
		$self->debug($c);
		if ($c ~~ ['+', '-', '*', '/']) {
			my $op = $c;
			$self->debug("op = $op");
			if ($self->{digit}) {
				push @expression, $self->{digit};
				$self->{digit} = '';
			}
			if (@$node) {
				$node = {$op => $node};
			} else {
				if ($op ~~ ['+', '-']) {
					my ($n, $idx) = $self->_parse_digit([@s[$i..$sl - 1]]);
					#push @$node, $n;
					$self->{digit} = $n;
					#push @expression, $self->{digit};
					push @$node, $self->{digit};
					$self->{digit} = '';
					$i = $i + $idx;
					
					#die $i;
					next;
					
				}
			}
			#push @$node, {$op => [@expression]};
			
			my $j;
			my $buff = '';
			my $cnt = 0;
			
			for ($j = $i + 1; $j < @s ; $j ++) {
				my $o = $s[$j];
				if ($o eq '(') {
					$cnt ++;
					#next;
				}
				if ($o eq ')') {
					$cnt --;
				} 
				
				
				
				$buff .= $o;
				$self->debug("$cnt, $buff");
				
				if ($cnt < 0 || $j == ($sl-1)) {
					#die $buff;
					#push @{$node->[-1]->{$op}}, $self->_parse ($buff);
					push @{$node->{$op}}, $self->_parse ($buff);
					last;
				} 
			}
			$i = $j;
			
			#@expression = ();
		} elsif ($c eq '(') {
			my $j;
			my $buff = '';
			for ($j = $i + 0; $j < @s ; $j ++) {
				my $o = $s[$j];
				$self->debug($o);
				$buff .= $o;
				my $cnt = 0;
				if ($o eq '(') {
					$cnt ++;
					next;
				}
				if ($o eq ')' || $j == scalar (@s) - 1) {
					$self->debug("$cnt , cnt00");
					$cnt --;
					if ($cnt == -1) {
						$self->debug("Buff sub = $buff");
						#die $function;
						push @$node, $self->_parse(substr($buff,1));
						#$node->{$function} = $self->_parse($buff);
						last;
					}
				}
			}
			$i = $j;
			$function = '';
			next;
		} elsif ($c =~ /[a-z]/) { # fuction processiong
			my $j;
			my $func_end = 0;
			my $buff = '';
			
			for ($j = $i; $j < @s ; $j ++) {
				
				my $o = $s[$j];
				if ($o eq '(') {
					$func_end = 1;
					my $j2;
					
					my $cnt = 1;
					for ($j2 = $j + 1; $j2 < @s ; $j2 ++) {
						my $o = $s[$j2];
						$buff .= $o;
						if ($o eq '(') {
							$cnt ++;
							next;
						}
						if ($o eq ')') {
							$cnt --;
							if ($cnt == 0) {
								last;
							}
						}
						
					}
					#$buff =~ s/^(.+).$/$1/;
					warn $buff;
					$j = $j2;
					last;
				} elsif ($func_end == 0) {
					$func .= $o;
				}
				
				
			}
			$i = $j;	
			
			$self->debug("Found function = $func, buff = $buff");
			push @$node, {$func => $self->_parse($buff)};
		} elsif ($c =~ /[0-9\.]/) {
			my ($n, $idx) = $self->_parse_digit([@s[$i..$sl - 1]]);
			$i = $idx + $i;
			push @$node, $n;
		}
	}
		
	return $node;
}

sub _parse_digit {
	my $self = shift;
	my $s = shift;
	my @s = @$s;
	
	my $i = 0;
	my $c = $s[0];
	
	my $node = [];
	
	$self->{digit} .= $c;
	my $j;
	my $buff = $self->{digit};
	my $had_non_digit = 0;
	for ($j = $i + 1; $j < @s ; $j ++) {
		my $o = $s[$j];
		if ($o =~ /[0-9\.]/) {
			$buff .= $o;
		} else {
			$had_non_digit = 1;
			last;
		}
	}
	
	if ($had_non_digit) {
		$i = $j - 1;
		$self->{digit} = substr($buff, 0, length ($buff) );
		$self->debug("Found digit = ".$self->{digit});
		push @$node, $self->{digit};
	} else {
		$i = $j;
		$self->{digit} = $buff;
		$self->debug("Found digit = ".$self->{digit});
		push @$node, $self->{digit};
	}
	return ($node, $i);
}
